### ISSUE TITLE
train_on_batch fails with MirroredStrategy

### ISSUE THREAD

**Issue by Flamefire** [tensorflow/tensorflow#39270]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): RHEL 7.5
- TensorFlow installed from (source or binary): source
- TensorFlow version (use command below): 2.1.0
- Python version: 3.7.4
- Bazel version (if compiling from source): 0.29.1
- GCC/Compiler version (if compiling from source): 8.3.0
- CUDA/cuDNN version: 10.1
- GPU model and memory: GTX1080Ti

**Describe the current behavior**

When running `train_on_batch` on a model created under a distribution strategy such as MirroredStrategy the error `ValueError: `handle` is not available outside the replica context or a `tf.distribute.Strategy.update()` call.` is thrown

Putting the `train_ob_batch` into the strategy scope does not change that error. However when doing so with a model containing a batch normalization layer (e.g. Resnet50) it throws a different error: `RuntimeError: `add_update` was called in a cross-replica context. This is not expected. If you require this feature, please file an issue.` (the other error location is not reached in that case as it bails out with the above)

**Describe the expected behavior**

`train_on_batch` should work

**Standalone code to reproduce the issue**
```
import tensorflow as tf

strategy = tf.distribute.MirroredStrategy()

with strategy.scope():
    (x_train, y_train), (x_test, y_test) = tf.keras.datasets.mnist.load_data()
    x_train, x_test = x_train / 255.0, x_test / 255.0
    train_dataset = tf.data.Dataset.from_tensor_slices((x_train, y_train)).batch(32)

    model = tf.keras.models.Sequential([
        tf.keras.layers.Flatten(input_shape=(28, 28)),
        tf.keras.layers.Dense(128, activation='relu'),
        tf.keras.layers.Dense(10),
    ])

    model.compile(
        optimizer=tf.keras.optimizers.SGD(),
        loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
        metrics=['accuracy'])

for x, y in train_dataset:
    model.train_on_batch(x, y)
```

**Other info / logs** Include any logs or source code that would be helpful to
Full traceback:
```
Traceback (most recent call last):
  File "tf_issue_train_on_batch.py", line 24, in <module>
    model.train_on_batch(x, y)
  File "/sw/installed/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/keras/engine/training.py", line 1078, in train_on_batch
    standalone=True)
  File "/sw/installed/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/keras/engine/training_v2_utils.py", line 433, in train_on_batch
    output_loss_metrics=model._output_loss_metrics)
  File "/sw/installed/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/eager/def_function.py", line 568, in __call__
    result = self._call(*args, **kwds)
  File "/sw/installed/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/eager/def_function.py", line 615, in _call
    self._initialize(args, kwds, add_initializers_to=initializers)
  File "/sw/installed/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/eager/def_function.py", line 497, in _initialize
    *args, **kwds))
  File "/sw/installed/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/eager/function.py", line 2389, in _get_concrete_function_internal_garbage_collected
    graph_function, _, _ = self._maybe_define_function(args, kwargs)
  File "/sw/installed/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/eager/function.py", line 2703, in _maybe_define_function
    graph_function = self._create_graph_function(args, kwargs)
  File "/sw/installed/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/eager/function.py", line 2593, in _create_graph_function
    capture_by_value=self._capture_by_value),
  File "/sw/installed/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/framework/func_graph.py", line 978, in func_graph_from_py_func
    func_outputs = python_func(*func_args, **func_kwargs)
  File "/sw/installed/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/eager/def_function.py", line 439, in wrapped_fn
    return weak_wrapped_fn().__wrapped__(*args, **kwds)
  File "/sw/installed/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/framework/func_graph.py", line 968, in wrapper
    raise e.ag_error_metadata.to_exception(e)
ValueError: in converted code:

    /sw/installed/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/keras/engine/training_eager.py:305 train_on_batch  *
        outs, total_loss, output_losses, masks = (
    /sw/installed/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/keras/engine/training_eager.py:273 _process_single_batch
        model.optimizer.apply_gradients(zip(grads, trainable_weights))
    /sw/installed/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/keras/optimizer_v2/optimizer_v2.py:444 apply_gradients
        kwargs={"name": name})
    /sw/installed/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/distribute/distribute_lib.py:1949 merge_call
        return self._merge_call(merge_fn, args, kwargs)
    /sw/installed/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/distribute/distribute_lib.py:1956 _merge_call
        return merge_fn(self._strategy, *args, **kwargs)
    /sw/installed/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/keras/optimizer_v2/optimizer_v2.py:485 _distributed_apply
        scope_name), distribution.extended.colocate_vars_with(var):
    /sw/installed/Python/3.7.4-GCCcore-8.3.0/lib/python3.7/contextlib.py:112 __enter__
        return next(self.gen)
    /sw/installed/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/framework/ops.py:4112 _colocate_with_for_gradient
        with self.colocate_with(op, ignore_existing):
    /sw/installed/Python/3.7.4-GCCcore-8.3.0/lib/python3.7/contextlib.py:112 __enter__
        return next(self.gen)
    /sw/installed/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/framework/ops.py:4161 colocate_with
        op = _op_to_colocate_with(op, self)
    /sw/installed/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/framework/ops.py:6548 _op_to_colocate_with
        if hasattr(v, "handle") and isinstance(v.handlTraceback (most recent call last):
  File "tf_issue_train_on_batch.py", line 24, in <module>
    model.train_on_batch(x, y)
  File "/sw/installed/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/keras/engine/training.py", line 1078, in train_on_batch
    standalone=True)
  File "/sw/installed/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/keras/engine/training_v2_utils.py", line 433, in train_on_batch
    output_loss_metrics=model._output_loss_metrics)
  File "/sw/installed/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/eager/def_function.py", line 568, in __call__
    result = self._call(*args, **kwds)
  File "/sw/installed/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/eager/def_function.py", line 615, in _call
    self._initialize(args, kwds, add_initializers_to=initializers)
  File "/sw/installed/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/eager/def_function.py", line 497, in _initialize
    *args, **kwds))
  File "/sw/installed/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/eager/function.py", line 2389, in _get_concrete_function_internal_garbage_collected
    graph_function, _, _ = self._maybe_define_function(args, kwargs)
  File "/sw/installed/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/eager/function.py", line 2703, in _maybe_define_function
    graph_function = self._create_graph_function(args, kwargs)
  File "/sw/installed/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/eager/function.py", line 2593, in _create_graph_function
    capture_by_value=self._capture_by_value),
  File "/sw/installed/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/framework/func_graph.py", line 978, in func_graph_from_py_func
    func_outputs = python_func(*func_args, **func_kwargs)
  File "/sw/installed/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/eager/def_function.py", line 439, in wrapped_fn
    return weak_wrapped_fn().__wrapped__(*args, **kwds)
  File "/sw/installed/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/framework/func_graph.py", line 968, in wrapper
    raise e.ag_error_metadata.to_exception(e)
ValueError: in converted code:

    /sw/installed/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/keras/engine/training_eager.py:305 train_on_batch  *
        outs, total_loss, output_losses, masks = (
    /sw/installed/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/keras/engine/training_eager.py:273 _process_single_batch
        model.optimizer.apply_gradients(zip(grads, trainable_weights))
    /sw/installed/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/keras/optimizer_v2/optimizer_v2.py:444 apply_gradients
        kwargs={"name": name})
    /sw/installed/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/distribute/distribute_lib.py:1949 merge_call
        return self._merge_call(merge_fn, args, kwargs)
    /sw/installed/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/distribute/distribute_lib.py:1956 _merge_call
        return merge_fn(self._strategy, *args, **kwargs)
    /sw/installed/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/keras/optimizer_v2/optimizer_v2.py:485 _distributed_apply
        scope_name), distribution.extended.colocate_vars_with(var):
    /sw/installed/Python/3.7.4-GCCcore-8.3.0/lib/python3.7/contextlib.py:112 __enter__
        return next(self.gen)
    /sw/installed/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/framework/ops.py:4112 _colocate_with_for_gradient
        with self.colocate_with(op, ignore_existing):
    /sw/installed/Python/3.7.4-GCCcore-8.3.0/lib/python3.7/contextlib.py:112 __enter__
        return next(self.gen)
    /sw/installed/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/framework/ops.py:4161 colocate_with
        op = _op_to_colocate_with(op, self)
    /sw/installed/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/framework/ops.py:6548 _op_to_colocate_with
        if hasattr(v, "handle") and isinstance(v.handle, Tensor):
    /sw/installed/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/distribute/values.py:720 handle
        raise ValueError("`handle` is not available outside the replica context"

    ValueError: `handle` is not available outside the replica context or a `tf.distribute.Strategy.update()` call.

```e, Tensor):
    /sw/installed/TensorFlow/2.1.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/tensorflow_core/python/distribute/values.py:720 handle
        raise ValueError("`handle` is not available outside the replica context"

    ValueError: `handle` is not available outside the replica context or a `tf.distribute.Strategy.update()` call.

**Comment by amahendrakar** [comment]
@Flamefire,
I was able to reproduce the error with [TF v2.1](https://colab.research.google.com/gist/amahendrakar/dd941f06afa3d22db01cfe186c87a6d6/39270-2-1.ipynb). However, the issue seems to be fixed with [TF v2.2.0-rc4](https://colab.research.google.com/gist/amahendrakar/dccbfd95695100d154b7bf34e0f6ca0d/39270.ipynb). Please find the attached gist. Thanks!

**Comment by Flamefire** [comment]
@amahendrakar I can confirm this. However I noticed a change in TF 2.2 to 2.1: Using `train_on_batch` in a tf.function is no longer possible as it errors out with `AttributeError: 'Tensor' object has no attribute 'numpy'`. Is `train_on_batch` supposed to be called outside a tf.function? My understanding was to put as much as possible into tf.functions.

This used to work in TF 2.1

See gist: https://colab.research.google.com/drive/1e3UwhKhQUFyYh5GsIO3TWOmvMynAPJUg?usp=sharing

**Comment by amahendrakar** [comment]
Was able to run the code with [TF v2.1](https://colab.research.google.com/gist/amahendrakar/4aa161d93375269ac26a929feca8f0b1/39270-2-1.ipynb). 

However, running the code with [TF v2.2](https://colab.research.google.com/gist/amahendrakar/5fbc906723702078352d3ea97464af9f/39270-2-2.ipynb#scrollTo=ZvOt6Ry6X7b-) throws an error stating `    AttributeError: 'Tensor' object has no attribute 'numpy'`. Please find the attached gist. Thanks!

**Comment by Flamefire** [comment]
As a note on that issue: It seems that it converts the Tensor (which is likely on GPU, isn't it?) to a numpy type (on CPU, AFAIK numpy types can't be on GPU), which involves a Device->Host copy and a synchronization. The value converted is the metric and loss values.

So doing that on every step seems wasteful to me as it may not be used at all. So it might be a good idea not to convert that eagerly into a numpy type which would resolve the AttributeError and maybe improve performance.

**Comment by goldiegadde** [comment]
@Flamefire the original issue of train_on_batch failed with MirroredStrategy is fixed in 2.2.0.

The second issue of Attribute Error when using train_on_batch with tf.function 
model.train_on_batch is disallowed inside a tf.function , since it is a high-level end point that manages its own tf.function since it creates an Iterator, converts results to NumPy, etc. The best thing to use is Model.train_step, which is just the logic of the training step and is safe to tf.function. 

Please let us know if that answers your questions and it is ok to close this issue.

**Comment by Flamefire** [comment]
Just for my understanding: In another issue I was told that `tf.function` is supposed to be the replacement for the TF 1 "graph mode" to recover (some of) the lost performance due to eager execution.   
However it now seems to be impossible to create a whole training loop or at least a single epoch iteration in "graph mode" (`tf.function`) as e.g. calling `train_on_batch` inside a `tf.function` is (now?) disallowed. So this still looses quite some performance especially because of the (potentially unneccessary) conversion to numpy types after each batch. To my understanding feeding the GPU new batches as fast as possible without first waiting on the result of previous computations is vital in leveraging the full performance of the GPU.

My point in using `train_on_batch` over `model.train_step` (besides the latter didn't exist in 2.1) was to remove as much overhead as possible as I figure `model.fit` does quite some work so it works with any model and any kind of inputs and outputs.

So bottom line: Shouldn't also "high level functions" like `rain_on_batch` be able to be put in `tf.function` so at least parts of the overhead are removed? I suppose quite some checks can be effectively removed when that functions is compiled for a given model and dataset type, and not going through all those Python abstractions but executing (compiled) binary code instead would be a win.

**Comment by saikumarchalla** [comment]
@Flamefire I executed the code in TF 2.5 and didnt face any errors. Could you please confirm and let us know if we can close the issue.

**Comment by Flamefire** [comment]
Thanks for the update. Well there is at least an open question:

>  Shouldn't also "high level functions" like train_on_batch be able to be put in tf.function so at least parts of the overhead are removed? I suppose quite some checks can be effectively removed when that functions is compiled for a given model and dataset type, and not going through all those Python abstractions but executing (compiled) binary code instead would be a win.

See my previous message for details.
